File: app/src/main/java/com/antyzero/smoksmog/ui/view/ViewPagerIndicator.java
package com.antyzero.smoksmog.ui.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.support.v4.view.ViewPager;
import android.util.AttributeSet;
import android.view.View;

import com.antyzero.smoksmog.R;

import java.util.ArrayList;
import java.util.List;


/**
 *
 */
public class ViewPagerIndicator extends View implements ViewPager.OnPageChangeListener {

    private List<Long> stationIds = new ArrayList<>();

    private int itemSpacing = 0;
    private int itemSideSize = 0;
    private int itemSideHalfSize = 0;

    private int desiredWidth = 0;
    private int desiredHeight = 0;

    private Paint paintActiveIndicator;
    private Paint paintIndicator;

    private Path arrow;

    private int activePosition = 0;

    public ViewPagerIndicator( Context context ) {
        super( context );
        init( context );
    }

    public ViewPagerIndicator( Context context, AttributeSet attrs ) {
        super( context, attrs );
        init( context );
    }

    public ViewPagerIndicator( Context context, AttributeSet attrs, int defStyleAttr ) {
        super( context, attrs, defStyleAttr );
        init( context );
    }

    private void init( Context context ) {
        itemSpacing = context.getResources().getDimensionPixelSize( R.dimen.indicator_item_spacing );
        itemSideSize = context.getResources().getDimensionPixelSize( R.dimen.indicator_item_size );
        itemSideHalfSize = itemSideSize / 2;

        paintActiveIndicator = new Paint();
        paintActiveIndicator.setColor( getResources().getColor( R.color.iron ) );
        paintActiveIndicator.setStyle( Paint.Style.FILL );
        paintActiveIndicator.setStrokeWidth( 0 );
        paintActiveIndicator.setAntiAlias( true );

        paintIndicator = new Paint( paintActiveIndicator );
        paintIndicator.setAlpha( 127 );

        arrow = new Path();
        arrow.reset();
        arrow.moveTo( itemSideHalfSize, itemSideHalfSize );
        arrow.lineTo( itemSideHalfSize, itemSideSize );
        arrow.lineTo( itemSideSize, 0 );
        arrow.lineTo( 0, itemSideHalfSize );
        arrow.close();
    }

    public void setStationIds( List<Long> stationIds ) {
        this.stationIds = stationIds;
        requestLayout();
        invalidate();
    }

    @Override
    protected void onMeasure( int widthMeasureSpec, int heightMeasureSpec ) {

        desiredHeight = 0;
        desiredWidth = 0;

        if ( !stationIds.isEmpty() ) {
            desiredHeight = itemSideSize;
            desiredWidth = itemSideSize;
            for ( int i = 1; i < stationIds.size(); i++ ) {
                desiredWidth += itemSideSize + itemSpacing;
            }
        }

        setMeasuredDimension( desiredWidth, desiredHeight );
    }

    @Override
    protected void onDraw( Canvas canvas ) {
        super.onDraw( canvas );
        for ( int i = 0; i < stationIds.size(); i++ ) {
            Paint paint = paintIndicator;

            if ( i == activePosition ) {
                paint = paintActiveIndicator;
            }

            if ( stationIds.get( i ) <= 0 ) {
                canvas.save();
                canvas.translate( ( itemSideSize + itemSpacing ) * i, 0 );
                canvas.drawPath( arrow, paint );
                canvas.restore();
            } else {
                int cx = itemSideHalfSize + ( itemSideSize + itemSpacing ) * i;
                int cy = this.itemSideHalfSize;
                canvas.drawCircle( cx, cy, this.itemSideHalfSize, paint );
            }
        }
    }

    @Override
    public void onPageScrolled( int position, float positionOffset, int positionOffsetPixels ) {

    }

    @Override
    public void onPageSelected( int position ) {
        if ( position != activePosition ) {
            activePosition = position;
            postInvalidate();
            // TODO we may ask to redraw only specific areas ?
        }
    }

    @Override
    public void onPageScrollStateChanged( int state ) {

    }
}

File: app/src/main/kotlin/com/antyzero/smoksmog/ui/view/ViewPagerIndicator.kt
package com.antyzero.smoksmog.ui.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.support.v4.view.ViewPager
import android.util.AttributeSet
import android.view.View
import com.antyzero.smoksmog.R
import com.antyzero.smoksmog.dsl.getCompatColor
import java.util.*


/**

 */
class ViewPagerIndicator @JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : View(context, attrs, defStyleAttr), ViewPager.OnPageChangeListener {

    private var stationIds: List<Long> = ArrayList()

    private var itemSpacing = 0
    private var itemSideSize = 0
    private var itemSideHalfSize = 0

    private var desiredWidth = 0
    private var desiredHeight = 0

    private var arrow: Path = Path()
    private var paintActiveIndicator: Paint = Paint()
    private var paintIndicator: Paint

    private var activePosition = 0

    init {

        itemSpacing = context.resources.getDimensionPixelSize(R.dimen.indicator_item_spacing)
        itemSideSize = context.resources.getDimensionPixelSize(R.dimen.indicator_item_size)
        itemSideHalfSize = itemSideSize / 2

        paintActiveIndicator.color = context.getCompatColor(R.color.iron)
        paintActiveIndicator.style = Paint.Style.FILL
        paintActiveIndicator.strokeWidth = 0f
        paintActiveIndicator.isAntiAlias = true

        paintIndicator = Paint(paintActiveIndicator)
        paintIndicator.alpha = 127

        arrow.reset()
        arrow.moveTo(itemSideHalfSize.toFloat(), itemSideHalfSize.toFloat())
        arrow.lineTo(itemSideHalfSize.toFloat(), itemSideSize.toFloat())
        arrow.lineTo(itemSideSize.toFloat(), 0f)
        arrow.lineTo(0f, itemSideHalfSize.toFloat())
        arrow.close()
    }

    fun setStationIds(stationIds: List<Long>) {
        this.stationIds = stationIds
        requestLayout()
        invalidate()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {

        desiredHeight = 0
        desiredWidth = 0

        if (!stationIds.isEmpty()) {
            desiredHeight = itemSideSize
            desiredWidth = itemSideSize
            for (i in 1..stationIds.size - 1) {
                desiredWidth += itemSideSize + itemSpacing
            }
        }

        setMeasuredDimension(desiredWidth, desiredHeight)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        for (i in stationIds.indices) {
            var paint: Paint = paintIndicator

            if (i == activePosition) {
                paint = paintActiveIndicator
            }

            if (stationIds[i] <= 0) {
                canvas.save()
                canvas.translate(((itemSideSize + itemSpacing) * i).toFloat(), 0f)
                canvas.drawPath(arrow!!, paint)
                canvas.restore()
            } else {
                val cx = itemSideHalfSize + (itemSideSize + itemSpacing) * i
                val cy = this.itemSideHalfSize
                canvas.drawCircle(cx.toFloat(), cy.toFloat(), this.itemSideHalfSize.toFloat(), paint)
            }
        }
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {

    }

    override fun onPageSelected(position: Int) {
        if (position != activePosition) {
            activePosition = position
            postInvalidate()
            // TODO we may ask to redraw only specific areas ?
        }
    }

    override fun onPageScrollStateChanged(state: Int) {

    }
}
